### PR TITLE
[REVIEW] Refactor concatenation logic and fix various bugs

### PR DESCRIPTION
**Pull request by vyasr**
This PR rewrites much of the internals of `cudf.concat`. It includes the following feature improvements and bug fixes:
1. Bug: When passed a single `cudf.Series` object with `ignore_index=True`, `cudf.concat` would upcast the output to a `DataFrame`.
2. Bug: When initializing a `cudf.DataFrame` using a `dict` _and_ providing column names, if none of the column names are actually in the dictionary the resulting `DataFrame` would be placed in an invalid state where the data would contain empty rows of size 0 but the index would have a size corresponding to the provided data columns. `pandas` returns a `DataFrame` with the expected shape (number of rows corresponding to the size of each provided data column, number of columns corresponding to the number specified in the column argument list), except all the data is filled with NaN. `cudf` agrees with `pandas` as long as at least one valid column exists in the data, it just needed to be modified for the case where no valid columns were provided. **EDIT**: Actually, the behavior is a bit more complicated than I originally stated. The number of rows in the resulting DataFrame depends on the index provided, so it should be empty when no index is given but have rows corresponding to `len(index)` when there is an index.
3. Feature: Column sorting was disabled in this function due to #6821. Now that bug has been resolved, sorting can be reenabled.
4. Bug: Various places were using the check `if columns`, which is invalid when the input is an `Index`.
5. Feature/Bug: Previously, concatenations based on the "inner" pattern would always sort the input columns simply by choice of algorithm. `pandas` does not do this, so we should be matching that behavior.

In addition, this PR includes the following changes to improve code quality and clarity:
1. Moves the definition of `Frame._concat` to `DataFrame`. The other `Frame` subclasses (`Index` and `Series`) each provide their own implementations, so the "default" implementation is really the `DataFrame` implementation.
2. Use Python `set` operations for identifying column intersections rather than a mix of `numpy` and `pandas` functions.
3. Various other minor changes to simplify parts of the code and reduce overhead from unnecessary operations like dictionary copies or complex iterations.

### CODE REVIEW

File: python/cudf/cudf/core/reshape.py [hunk 1]
```diff
@@ -275,44 +277,50 @@ def concat(objs, axis=0, join="outer", ignore_index=False, sort=None):
...
         if empty_inner:
             # if join is inner and it contains an empty df
             # we return an empty df
             return df.head(0)
+
+        # This check uses `sort is not False` rather than just `sort=True`
+        # to differentiate between a user-provided `False` value and the
+        # default `None`. This is necessary for pandas compatibility, even
+        # though `True` and `False` are the only valid options from the user.
         if not match_index and sort is not False:
```
Comment (isVoid) [inline python/cudf/cudf/core/reshape.py:321]: I think `sort` is by default `False` now? https://pandas.pydata.org/docs/reference/api/pandas.concat.html . Note if we are changing ours, it should be a different PR.

File: python/cudf/cudf/tests/test_concat.py [hunk 2]
```diff
@@ -1203,3 +1203,14 @@ def test_concat_join_empty_dataframes_axis_1(
     assert_eq(
         expected, actual, check_index_type=False, check_column_type=False
     )
+
+
+def test_concat_preserve_order():
```
Comment (isVoid) [inline python/cudf/cudf/tests/test_concat.py:1208]: Do we need more tests/revisal of previous tests to cover the bugs fixed in this PR?

**Review by isVoid — COMMENTED** [review]
Assuming `Frame._concat` -> `DataFrame._concat` is just trivial code move. Other refactors look pretty good. A few nitpicks/QST below.

Side note - would be great to split a large PR like this into several smaller ones in the future.

File: python/cudf/cudf/core/reshape.py (hunk 1 above)
Comment (vyasr) [reply python/cudf/cudf/core/reshape.py:321]: Was the default for pandas concat `sort=True` at one point?

File: python/cudf/cudf/tests/test_concat.py (hunk 2 above)
Comment (vyasr) [reply python/cudf/cudf/tests/test_concat.py:1208]: Bugs 1 and 2 are both already detected by other test code paths now that 3 and 5 are fixed (this new test is for 5). Bug 4 is more an implementation detail than an API change, so I didn't bother to add one. I can not be lazy and create a test for that case (using an `Index` object to specify the columns) if you think it's worth it.

**Comment by vyasr** [comment]
> Assuming `Frame._concat` -> `DataFrame._concat` is just trivial code move. Other refactors look pretty good. A few nitpicks/QST below.
> 
> Side note - would be great to split a large PR like this into several smaller ones in the future.

Yeah, sorry about that. I was originally trying for a much smaller PR, but the changes snowballed since each change I made revealed additional bugs or code smells. This already conflicted with a number of my other open PRs so I didn't try and split this one up any further. I did consider undoing the `Frame->DataFrame` move since that certainly makes it difficult to review. If you'd prefer, I can move the code back for this PR review so that you can more easily see the actually meaningful changes, then make a subsequent PR with just the move.

File: python/cudf/cudf/core/reshape.py (hunk 1 above)
Comment (isVoid) [reply python/cudf/cudf/core/reshape.py:321]: Prior to v1.0 it was defaulted at `None`, but the meaning of `sort=None` is actually `sort=True`. They changed that after v1.0: https://github.com/pandas-dev/pandas/blob/13653812b43aab0d4be29177f9c849c07414a7bc/pandas/core/reshape/concat.py#L86-L89

**Comment by isVoid** [comment]
I think it should be ok for this PR to keep as it is.

File: python/cudf/cudf/tests/test_concat.py (hunk 2 above)
Comment (isVoid) [reply python/cudf/cudf/tests/test_concat.py:1208]: We shouldn't add tests for implementation detail though.

File: python/cudf/cudf/core/reshape.py (hunk 1 above)
Comment (isVoid) [reply python/cudf/cudf/core/reshape.py:321]: Sounds like we are missing a test for `None` parameter for `sort`?

Comment (vyasr) [reply python/cudf/cudf/core/reshape.py:321]: Got it. So the current behavior of `cudf` is to have `sort=None`, which means `sort=False`, so that's consistent with `pandas`. I assumed (but I don't remember whether I tested) that the line of code we're currently discussing (the `sort is not False`) is specifically to handle an edge case where `pandas` sorted the data even when `sort=False`, but I'm not 100% sure if that's the case. I'll double-check that before finalizing this PR, maybe we can simply change the behavior in this line. I know that it caused some tests to fail, but maybe those tests should be modified.

File: python/cudf/cudf/tests/test_concat.py (hunk 2 above)
Comment (vyasr) [reply python/cudf/cudf/tests/test_concat.py:1208]: Right, I agree. It's just that this implementation detail has an API implication. Users can call `pd.concat(..., columns=pd.Index(list_of_column_names))`, which should be equivalent to `pd.concat(..., columns=list_of_column_names)`. However, without the fix I did for bug 4, the first case will error. Checking `if []` is valid Python, it will evaluate to `False` if the list is empty and `True` otherwise, but `pd.Index` doesn't support the same check because it behaves more like `numpy.ndarray` where you have to do elementwise true/false checks or something like `all(index == 0)`. Therefore, we can't just do `if columns`, we have to do `if len(columns) != 0`. The relevant test to add would be to do `pd.concat` with a `pd.Index` or `cudf.Index` as the `columns` object. I'll just not be lazy and add the test :)

File: python/cudf/cudf/core/reshape.py (hunk 1 above)
Comment (vyasr) [reply python/cudf/cudf/core/reshape.py:321]: Yes, this is necessary. I think it's specifically an issue for non-integral indexes where `cudf` won't generate a sorted index without explicitly calling `sort_index` here but pandas will sort on the index even when the user doesn't request a sort. See `test_concat.py::test_pandas_concat_compatibility_axis1`.

**Comment by vyasr** [comment]
@isVoid especially with the merge of some other PRs upstream this diff became pretty unwieldy, so I moved `_concat` back to `Frame` from `DataFrame` and reshuffled methods around to avoid showing meaningless diffs. I'll have to make a follow-up PR for some minor changes associated with #7876, so I'll toss the `Frame._concat`->`DataFrame._concat` move into that PR as well. I also added tests for all of the cases that I was implicitly relying on other tests for. This PR should be ready for a final review now.

File: python/cudf/cudf/core/dataframe.py [hunk 3]
```diff
@@ -427,38 +427,46 @@ def _init_from_list_like(self, data, index=None, columns=None):
...
+                # we need to create an empty DataFrame. To match pandas, the
+                # size of the dataframe must match the provided index, so we
+                # need to return a masked array of nulls if an index is given.
+                row_count = 0 if index is None else len(index)
+                masked = index is not None
                 data = {
-                    key: cudf.core.column.column_empty(row_count=0, dtype=None)
+                    key: cudf.core.column.column_empty(
+                        row_count=row_count, dtype=None, masked=masked,
+                    )
```
Comment (brandon-b-miller) [inline python/cudf/cudf/core/dataframe.py:462]: These columns are all the exact same right? Is `column_empty` faster than just making a single masked column and then doing `len(extra_cols)` deep copies of that original column?

shame we can't allocate all of these at once somehow.

File: python/cudf/cudf/core/reshape.py [hunk 4]
```diff
@@ -201,44 +203,46 @@ def concat(objs, axis=0, join="outer", ignore_index=False, sort=None):
     1      b       2  monkey  george
     """
 
+    # TODO: Do we really need to have different error messages for an empty
+    # list and a list of None?
```
Comment (brandon-b-miller) [inline python/cudf/cudf/core/reshape.py:207]: The answer to this is probably no but  we usually to just follow whatever pandas does. 
```
>>> pd.concat([])
ValueError: No objects to concatenate
>>> pd.concat([None, None, None])
ValueError: All objects passed were None
```

**Review by brandon-b-miller — COMMENTED** [review]
Looks like this tackles a lot of little things that have been hanging around in the concat workflow that could be simpler or more efficient. Had a few small questions.

**Review by brandon-b-miller — APPROVED** [review]

**Review by isVoid — COMMENTED** [review]
mostly lgtm, with 1 small question

**Comment by vyasr** [comment]
rerun tests

File: python/cudf/cudf/core/dataframe.py [hunk 5]
```diff
@@ -427,38 +427,44 @@ def _init_from_list_like(self, data, index=None, columns=None):
...
+                # If keys is empty, none of the data keys match the columns, so
+                # we need to create an empty DataFrame. To match pandas, the
+                # size of the dataframe must match the provided index, so we
+                # need to return a masked array of nulls if an index is given.
+                row_count = 0 if index is None else len(index)
+                masked = index is not None
+                empty_column = cudf.core.column.column_empty(
+                    row_count=row_count, dtype=None, masked=masked,
+                )
+                data = {k: empty_column.copy(deep=True) for k in extra_cols}
```
[the remark this answers is not in the document]
Comment (vyasr) [reply python/cudf/cudf/core/dataframe.py]: Closing the loop here: @brandon-b-miller suggested that using a copy here might be faster in a previous comment. I incorrectly applied this patch by forgetting to remove the extra column created in the approach here. However, I also benchmarked and found that the constructor is actually faster than the deep copy anyway, so I switched back.

**Review by isVoid — APPROVED** [review]

**Comment by vyasr** [comment]
@gpucibot merge
